nntrainer/layers/layer_node.h

```diff
@@ -52,6 +52,18 @@ class Activation;
 class SharedFrom;
 } // namespace props
 
+/**
+ * @brief Enum class for the various types of inplace modes supported by layer
+ *
+ */
+enum class InPlace {
+  NONE,           /**< layer is not inplace */
+  RESTRICTING,    /**< layer is in-place and does place restriction on layers
```

Comment:
how about PREEMPTIVE to convey meaning?

Reply:
I don't think that the meaning of preemptive applies here. Check https://github.com/nnstreamer/nntrainer/pull/1689#discussion_r746186559 for more details.

Reply:
Cool.

Reply:
Idea to refine this.

1. There is 2 properties of requestSpec()
 - clean inplace and dirty inplace
 - if clean (meaning that tensor has first set and not overwritten) it is allowed to inplace (clean or dirty)
 - if dirty (meaning that tensor was overwritten) it is allowed to clean inplace but cannot do dirty inplace

| current tensor state | trying tensor state | can in place? |
|-----------------------|---------------------|------------|
| clean                        | clean inplace         | O |
| clean | dirty inplace | O |
| dirty | clean inplace | O |
| dirty | dirty inplace | X |

Reply:
Yes, clean and dirty names sound good, but have a slightly different concepts. `clean` and `dirty` seem to give the notion if the layer operation will modify the input tensor. But instead, there are two concepts:
- if the layer modifies the input tensor
- does the layer require its input/output tensor in its backward operation

So, given a state associated to the output from the previous layer output, how does the current state modify its input with inplace operation (1) and what is the requirement of the current layer wrt to its backward operation (2), both needs to be represented. Lets define two concepts:
a. Layer operation : modifies/not-modifies
b. Layer backwarding requirement : needed/non-needed

Based on (b), an output of a layer can be assumed to be needed/non-needed. It will be non-needed for out of place layers. If the inputs for a layer not-needed (b), then this layer operation can be always performed in-place (if possible) (a) and the tensor state will be updated based on (b).  If the inputs for a layer are needed, then only not-modifies (b) layer operation can work in-place, independent of layer backwarding requirement. If the layer operation is modifies (b), then it will have to work out-of-place.

As in-place optimization is performed at the compile time, tensor objects are not yet created, and its better to associate these properties to layer (defining the way they are operating) and can be checked with its neighbors based on their connections.

PS. there can be better names than `modifies` and `needed`.

Reply:
having `modified`, `needed` sounds good, for needed we won't need another boolean but we can just check the max execution order is less then new min execution order.

Right, I am thinking to extend the concept to the normal tensor (not limited to the inputs, outputs).

| previously modified | needed after my execution order | view type | actually requested |
|-|-|-|-|
| - | - | READ_ONLY_VIEW | `view()` |
| x | x | MAYBE_MODIFIYING_VIEW | `view()` |
| o | o | MAYBE_MODIFIYING_VIEW |  `request()` |
| o | x | MAYBE_MODIFIYING_VIEW |  `view()` |
| x | o | MAYBE_MODIFIYING_VIEW |  `request()` |

Which contracts to only checking `needed after my execution order`

In this way, finalize does not need to query execute inplace, nor graph need to check whether it should inplace optimize but we have defined rule to hint graph how it should be initialized.

Any thoughts?

fyi) please check out #1700 as a proposal.

Reply:
I prefer the modified and needed rather than clean and dirty. I agree it is different. I wonder requrest() means create()??